### ISSUE TITLE
bug(mac): Java does not appear to be 100% compatible with InputMethodKit

### ISSUE THREAD

**Issue by mcdurdin** [keymanapp/keyman#3935]
This is a container issue for compatibility issues with Java on macOS.

The following may be related: https://youtrack.jetbrains.com/issue/JBR-1758

Related issues:
* #174 [Mac] Keyman Mac does not work with Oxygen XML editor
* #458 [Mac] JDK compatibility problem (affects OmegaT)
* #3243 Keyman not compatible with ELAN

**Comment by mcdurdin** [comment]
See http://bugs.java.com/bugdatabase/view_bug.do?bug_id=JDK-8257650 for our submitted bug.

**Comment by mcdurdin** [comment]
I've gone ahead and closed #174, #458 and #3243, to reduce duplication as they all pertain to the same basic problem.

Note that there may be additional information in those issues that we want to refer to later.

**Comment by Matt-SIL** [comment]
Have there been any developments with this issue? I just posted on the ELAN forum before seeing this. 

https://archive.mpi.nl/forums/t/trouble-with-input-from-syllabic-orthography-keyboard/3468

**Comment by mcdurdin** [comment]
We have not had any progress with this; it is a limitation in Java and we submitted an issue 4 years ago for it (https://bugs.java.com/bugdatabase/view_bug.do?bug_id=JDK-8195675). Feel free to submit another bug to Oracle!

**Comment by Matt-SIL** [comment]
Ok, I'm not sure I know enough to submit a bug with all the technical info they need but I would live to tell them this potentially affects a large number of minority languages.

**Comment by mcdurdin** [comment]
It's difficult -- I hope that bug reports from various people would help them to understand the scope. In theory we could submit a patch, but the amount of legwork that is required to do that seems daunting! (This seems to be a problem with quite a few large open source projects -- they develop policies and procedures which work for them but effectively lock out external contributors who already have to navigate huge technical complexity in order to submit patches, and the additional policy complexity just makes it too hard.)

**Comment by JuarezRC** [comment]
Hi @mcdurdin, I have a similar issue running the IPA(SIL) Keyboard on Mac. They keyboard works well everywhere, except for ELAN.  Any recommendation where I should look for a solution?

**Comment by mcdurdin** [comment]
@JuarezRC, yes, this is a long-standing issue. We are waiting for Java developers to fix things on their end -- Keyman is behaving correctly; Java (which ELAN is built on) is not. The two bug reports above have had no response from Oracle or any Java Development Kit developers; we may need to try and raise them in forums or Java dev mailing lists. We have not had the in-team capacity to chase these up until now.

Can I encourage you to also open a bug report or ask on the Java Discussion group (looks like awkward timing though as they are migrating their Discussion databases in the next couple of days -- may want to wait a week).

* Java Discussion Group: https://community.oracle.com/tech/developers/categories/jug-discuss
* Java Bug Report: https://bugs.java.com/bugdatabase/
* Java Developer Mailing List: https://mail.openjdk.org/mailman/listinfo/jdk-dev

Other possible workarounds, even if not palatable -- mentioned for completeness:
* Use ELAN on Windows, where AFAIK things work correctly.
* Copy and paste text into ELAN

We will also have another go at contributing a patch to JDK when we find the capacity (@sgschantz this may be something you could look into?)

**Comment by JuarezRC** [comment]
Thanks! I will follow up on this.

**Comment by mcdurdin** [comment]
OpenJDK:
* https://bugs.openjdk.org/browse/JDK-8195675
* https://bugs.openjdk.org/browse/JDK-8257650 (dup)

**Comment by mcdurdin** [comment]
@srl295 please review

**Comment by mcdurdin** [comment]
See https://github.com/keymanapp/keyman/pull/3944 for sample patch

**Comment by srl295** [comment]
PR: https://github.com/openjdk/jdk/pull/17921

**Comment by mcdurdin** [comment]
Some positive progress on this with openjdk, follow links above

**Comment by mcdurdin** [comment]
See also https://community.software.sil.org/t/keyman-and-xlingpaper/8580/1 (presumably XMLMind which XLingPaper is based on)

**Comment by AndyBlack** [comment]
> See also https://community.software.sil.org/t/keyman-and-xlingpaper/8580/1 (presumably XMLMind which XLingPaper is based on)

Yes, the XMLmind XL Editor (aka XXE) is written in Java and XLingPaper is a set of configuration files for it.  The challenge we'll face is that most XLingPaper users are using an older version of XXE which still had a personal use license.  The version they're using will not have the fix to the openjdk.  Sigh.

**Comment by mcdurdin** [comment]
@srl295 do you have any updates on the OpenJDK fix?

**Comment by srl295** [comment]
The OpenJDK PR is approved and should merge tomorrow into Java 23

**Comment by mcdurdin** [comment]
Process: we'll keep this open as we track back-ports into earlier versions of Java.

**Comment by AndyBlack** [comment]
If I understand this correctly, then you will try to make this needed change in earlier versions of Java.  Thank you!
For XLingPaper, it looks like the version of the XMLmind XML Editor we're using is based on Java version 1.8.0_181 (or higher).
For the other Java-based applications I'm responsible for, we've been using 1.8.0_162 (which is the last "free" one, I understand).  These applications  are https://software.sil.org/lingtree/, https://software.sil.org/lingtree/diagsap/, https://software.sil.org/paws/, and https://software.sil.org/asheninka/.

**Comment by srl295** [comment]
> If I understand this correctly, then you will try to make this needed change in earlier versions of Java.  Thank you!
> 
> For XLingPaper, it looks like the version of the XMLmind XML Editor we're using is based on Java version 1.8.0_181 (or higher).
> 
> For the other Java-based applications I'm responsible for, we've been using 1.8.0_162 (which is the last "free" one, I understand).  These applications  are https://software.sil.org/lingtree/, https://software.sil.org/lingtree/diagsap/, https://software.sil.org/paws/, and https://software.sil.org/asheninka/.

Ouch 1.8? I think 11 is the earliest in the openjdk tree.  1.8 may be challenging. 

As to "free" are you familiar with Temurin? https://adoptium.net/temurin/ fully certified binaries for v21.. free.

**Comment by AndyBlack** [comment]
I've been keeping with 8 mainly because it came with easy to use installer mechanisms for Windows, Mac, and Linux.  I have built one other Java app using 17 but for Windows only and found the learning curve for installers a bit high.  I'm not against climbing the learning curve, it's just appears to be very time consuming since there is so much I do not control...
For 17, I've used Azul Zulu which is also free.  They even fixed a bug I found...  See https://www.azul.com/downloads/?package=jdk#zulu

**Comment by srl295** [comment]
> I cannot control the version for the XMLmind XML Editor. For the apps I'm responsible for, I've been keeping with 8 mainly because it came with easy-to-use installer mechanisms for Windows, Mac, and Linux. I have built one other Java app using 17 but for Windows only and found the learning curve for installers a bit high. I'm not against climbing the learning curve, it just appears to be very time consuming since there is so much I do not control... Having some examples to follow would be helpful. 

Temurin has installers for Windows. Here's an article about using jlink with temurin. https://adoptium.net/blog/2021/10/jlink-to-produce-own-runtime/ 

jpackage can build installers https://docs.oracle.com/en/java/javase/17/jpackage/packaging-overview.html#GUID-E0966C49-ABBB-46A2-8DF7-1D3F96640F05

> For 17, I've used Azul Zulu which is also free. They even fixed a bug I found... See https://www.azul.com/downloads/?package=jdk#zulu

Azul says it is builds of openjdk. So if it were backported to OpenJDK 17, it should show up in Azul.

**Comment by AndyBlack** [comment]
Thanks

**Comment by srl295** [comment]
I've finally created a backport for v21 https://bugs.openjdk.org/browse/JDK-8342073 - I have a lot of questions about the process (it's literally been many years since I have done a backport for them) but it's in progress. If this goes well I'll replicate it back as far as I can.

**Comment by srl295** [comment]
Requested backports for:
- [ ] jdk8
- [ ] jdk11
- [ ] jdk17
- [ ] jdk21

**Comment by srl295** [comment]
Update^ All four of these have been tested on mac and found to fix the issue.

**Comment by AndyBlack** [comment]
Good news!
I'm not familiar with backports and how they work.  I'm currently using Azul Zulu's JDK 17 (17.0.11+9; its Bundle name is Zulu 17.50.19).  Does this mean that at some point Azul will include this fix for their Java 17 JDK and at that time I can get that version and then Keyman will work with my Java apps built using this?

**Comment by srl295** [comment]
Yes that’s what it means.

Would you like me to try to get a jdk 17 package to you to test ?

**Comment by srl295** [comment]
Ok 21 was merged! Others to follow.

**Comment by AndyBlack** [comment]
Yes, a jdk 17 package to test would be helpful.  Since I wrote earlier, I've learned how to use jpackage to create installers.

If it's not too much trouble, a version 8 jre would also be nice to test (this would be for the XLingPaper configuration files that are for the XMLmind XML Editor I mentioned. here back in April.  I could try swapping out the jre we're currently using with the new one and see if it runs as before, but also allows for Keyman keyboards.

**Comment by srl295** [comment]
@AndyBlack I sent you a gchat and also a share link to a directory w/ zips for 17 and 8.  I used "make images" to build it. Let me know if it works or not.

**Comment by AndyBlack** [comment]
@srl295 Thanks!  It may be a day or two before I can test these.

**Comment by srl295** [comment]
17 integrated

**Comment by mcdurdin** [comment]
@srl295 are there any further updates on jdk 8 and 11?

**Comment by AndyBlack** [comment]
As for JDK 23, it turns out that on macOS Java Swing applications work well with Keyman but JavaFX ones still do not.  I have two GitHub projects you can use to see this issue:

1. https://github.com/AndyBlack/KeymanSwing and
2. https://github.com/AndyBlack/KeymanFX.

I submitted an issue request with Azul on this to look into it.  On March 6, 2025 they said:

> We understand the importance of Keyman for multilingual users, and we'll investigate why this issue persists specifically with JavaFX on macOS, even though it's been addressed for Swing.
Our team will be reviewing the information you provided, including the references to the previous Keyman issue and your helpful GitHub projects illustrating the problem.
We'll focus on replicating the issue with the Hebrew (SIL) keyboard on macOS with Azul Zulu 23.32.11 and then begin troubleshooting.
Please allow us some time to investigate this further. We'll keep you updated on our progress and any potential solutions or workarounds we find.
Thank you for your patience and understanding.

**Comment by srl295** [comment]
11 was merged, still waiting on 8

**Comment by AndyBlack** [comment]
Here's what I've found about using Keyman with the XMLmind XML Editor (aka XX) on macOS:
XXE uses a DOM/Web page interface for the vast majority of the user input.  It also has a few places where one types into a text box.

Using Java 23 with the AWT fix with Keyman results in the correct result in a text box, but results in nothing at all happening in the web page portion.  One can type using Keyman into, say, TextEdit and then copy it into the web page portion of XXE.

Using the Java 8 that comes with XXE with Keyman results in the English letters showing in both the text box and the web page.  One can also copy and paste from another app.

On Windows, both the text box and the web page get the correct result using the Java 8 included with XXE.

XXE's documentation on its DocumentView includes the following (for what it's worth):
Typing a character in the XML editor works like this:

1. Let's say user types a newline character in a XHTML pre. This generates a KEY_TYPED event.
2. This event is bound to command insertControlCharOrSplit with parameter "\n".
3. This command uses the text location of the dot mark to find the view representing the pre.
4. The command queries the view to know if its accepts the newline character. For a pre (which has display: pre), this is the case.
5. The command inserts a '\n' character at text location given by the dot mark.
6. The document notifies its DocumentListeners that character '\n' has been inserted in a given text node (contained in the pre) at a given offset.
7. Among these DocumentListeners, there is the DocumentView. It finds the view which needs to be updated. In this case, it is a TextNodeView which represents the text contained in the pre. Then it invokes method charInserted which is an optimized way to request the view to refresh itself.
8. Among these DocumentListeners, there is also the MarkManager. It moves the dot mark by one character, then it notifies its MarkChangeListeners that dot has changed.
9. Among these MarkChangeListeners, there is the DocumentView. It finds the highlight representing the dot mark in order to refresh it. This highlight is called the caret. Then DocumentView requests the caret to relayout and repaint itself.

**Comment by darcywong00** [comment]
@srl295 fixed OpenJDK 8, 11, 17, 23
openjdk/jdk#17921

Since we're short-staffed, I think we have to push this to Future
